task_geo/data_sources/nasa/nasa.py

```diff
@@ -0,0 +1,214 @@
+                                        sub_region, lon, and lat
+        start_date(datetime64): Start date for the time series
+        end_date(pandas.Timestamp): End date fo rthe time series (optional)
+        parms(list of strings): Desired data, accepted are 'temperature',
+                                'humidity', and 'pressure' (optional)
+        level(string):  country, region, or sub_region, tells us at what
+                        granularity we want the data (eg for US sub_region, ie
+                        county, might make more sense than region, ie states,
+                        but in France region could be better, while for
+                        Luxembourg we probably want country) (optional)
+
+    Return:
+    ------
+        pandas.DataFrame:   Columns are country, region, sub_region (non-null),
+                            lon, lat, date, and the desired data.
+    """
+    df_locations = df_locations[
+        ~pd.isna(df_locations[[level, 'lon', 'lat']]).all(axis=1)
+    ]
+    location_data = ['country', 'region', 'sub_region', 'lon', 'lat']
+    locations = df_locations[location_data].drop_duplicates()
+
+    str_start_date = str(start_date.date()).replace('-', '')
+
+    if end_date is None:
+        str_end_date = str(pd.Timestamp.today().date()).replace('-', '')
+    else:
+        str_end_date = str(end_date.date()).replace('-', '')
+
+    all_parms = list(itertools.chain.from_iterable([PARAMETERS[p]
+                                                    for p in parms]))
+    parms_str = f"parameters={','.join(all_parms)}"
+
+    return pd.concat([
+        nasa_data_loc(row.country, row.region, row.sub_region, row.lat,
+                      row.lon, str_start_date, str_end_date, parms_str)
+        for row in locations.itertuples()
+    ])
+
+
+def nasa_formatter(df_nasa, parms):
+    """
+    Format the NASA data.
+
+    Parameters
+    ----------
+    df_nasa : pandas.DataFrame
+    parms : list of strings
+
+    Returns
+    -------
+    df_nasa : pandas.DataFrame
+
+    """
+    # date column
+    df_nasa.reset_index(inplace=True, drop=False)
+    df_nasa.rename(columns={'index': 'date'}, inplace=True)
+    df_nasa['date'] = df_nasa['date'].apply(lambda dt:
+                                            pd.to_datetime(
+                                                f"{dt[:4]}-{dt[4:6]}-{dt[6:]}"
+                                            ).date())
+
+    # reorder columns
+    base_parms = ['country', 'region', 'sub_region', 'lon', 'lat', 'date']
+    all_parms = list(itertools.chain.from_iterable([PARAMETERS[p]
+                                                    for p in parms]))
+    df_nasa = df_nasa[base_parms + all_parms]
+
+    # rename columns
+    df_nasa.rename(columns=COL_NAMES, inplace=True)
+
+    return df_nasa
+
+
+def nasa_meteo_data(df, start_date, end_date=None,
+                    parms=["temperature", "humidity", "pressure"],
```

Comment:
[Never use a list as a default parameter.](https://florimond.dev/blog/articles/2018/08/python-mutable-defaults-are-the-source-of-all-evil/) Use None, and set them inside the function

Reply:
Changed to None and defaulting for the keys of PARAMETERS.